### ISSUE TITLE
V3 Priorities

### ISSUE THREAD

**Issue by mikeshardmind** [mikeshardmind/SinbadCogs#57]
So this issue marks a development freeze on anything but bugfixes for any cogs written for Red (v2) in here. 

This is in preparation for getting everything ready for v3.

Some cogs need rewriting, others had features I knew would be easier to get working with the rewrite of discord.py that didn't get implemented [on v2] as a result.

Here's my current priority list (subject to change pending feedback from people who use my cogs)

- [X] BanSync, 
- [X] ServerWhitelist / ServerBlacklist
- [X] Autorooms & TempChannels
- [X] QuoteTools
- [X] ChannelRelays
- [X] AntiMentionSpam 
- [x]  Role Management System
- [X] Embedmaker
- [x] RSS

Future plans:

- [ ] SuggestionBox
- [ ] Announcer (pub/sub based)

Anything not listed here is not a priority as of now. 

Additionally, the following cogs are not being ported for various reasons:

advstatus (should be core feature in v3)
streamannouncer (This one might change, but realistically, I'd prefer to make the PR to red if needed)
audionotifier (Red's audio cog has this more than covered now)
___

Feedback about which cogs you'd like to see prioritized or what features you'd like to see are always welcome.

**Comment by HelmicNewciv** [comment]
I think ReportTool should be core.  I don't know if that'd mean something like merging ReportTool into ModLog as an optional setting (the command showing up to users only if it's enabled).  It's the sort of thing that I wish more servers would use, and any server that cares enough to maintain a mod log channel would benefit from users being able to discretely contact moderators.

**Comment by mikeshardmind** [comment]
@HelmicNewciv 
https://github.com/Cog-Creators/Red-DiscordBot/issues/1280
Awaiting response, This will exist either in Red, or as a cog for Red one way or another.

**Comment by mikeshardmind** [comment]
WIP: This [ReportTool] will be a core feature for Red https://github.com/Cog-Creators/Red-DiscordBot/pull/1281

**Comment by mikeshardmind** [comment]
Bansync is v3 ready now.

**Comment by mikeshardmind** [comment]
dd5fcbf removes the 1st item on that list,

# Update on remaining things:
## Autorooms & TempChannels
Will be implemented soon

## QuoteTools & ChannelRelays
Will be looking into a way to better handle a few things in these
This should be handled after AutoRooms, but possibly before TempChannels

## Announcer
There's an announcement tool in core red now, considering if this is a sufficient reason to drop mine or if I should instead keep it since the functionalities are different

## SuggestionBox
This is on the back burner as I still want to turn this into a more generic form tool, but have not decided the best means of this yet

## EmbedMaker
This is getting an overhaul. I'll be adding a way to get a quick embed for people who already know what they want, custom color support, embeds from markdown blocks, and a few other things.

**Comment by mikeshardmind** [comment]
These are currently all waiting on a set of possibly breaking changes in V3 before continuing.

**Comment by mikeshardmind** [comment]
Relay stuff is in and tested. Embed stuff is next unless I get motivated to do something else first instead.

**Comment by mikeshardmind** [comment]
@HelmicNewciv report tool is in latest beta release of V3 (as reports) with added functionality of moderator interaction. 

See: [red#1280](https://github.com/Cog-Creators/Red-DiscordBot/issues/1280)

**Comment by HelmicNewciv** [comment]
Think that was the last major roadblock before switching over to V3 for me.  Definitely look forward to being able to report stuff to moderators on more Discords.

**Comment by mikeshardmind** [comment]
Embedmaker is in with quite a bit of additional functionality, including timezone aware timestamps with some level of natural language parsing. `November 5th 2018 9pm EST` would be correctly parsed 

Relays are in

Autorooms, temp channels, and quotetools should also all be ready soon.

For anyone who was looking forward to the last project I sort of started, but never completed on v2 (the advanced role management system), It's a WIP for v3 now, the bare bones of which is in, but not ready for public use yet.

**Comment by netflixandbill** [comment]
any plans with the role systems? when it comes to the suggestions or announcer cogs, what if there was a formatted style that the owner or admins could create to have different fields for info organized with an embed and the bot would relay it to a channel? 
my idea with this is that the user would do a command like [p]review and the bot would send the user a DM much like the embed maker cog does where it asks them the questions/prompts/admin created fields and posts to #game-reviews.

[game title=]
[system=]
[time played=]
[rating/10=]
[review=]

**Comment by mikeshardmind** [comment]
@netflixandbill 
The role systems are still a heavy WIP. Current Ideas:

1. The main role management cog with have stickyrole support baked in. There's also just the standalone stickyroles cog for people who dont need the full suite
2. Roles can have required roles to be joined, and those can be a list of any, a list of all, or some combination, or a list of exclusions
3. Roles can be set to be self-removable (but default to not)
4. Roles can be joined/left using reactions or commands (so far, only emoji is supported, and only for joining)

Regarding the formatted style options for suggestionbox:

I like the idea, and this will be a part of it.

Regarding embeded Announcements:

I won't be re-implementing the logic from embedmaker, but I'll make it available at a later date to users who have both cogs to use the embedmaker to create their announcement.

**Comment by mikeshardmind** [comment]
QuoteTools and RoomTools are both in now. RoomTools is a merge of autorooms and tempchannels with menuing hidden for any disabled. Quotetools's logic has been simplified down to a single command, while retaining most of the prior functionality. More to be added on this at a later date.

**Comment by mikeshardmind** [comment]
RSS & Announcer Soon(TM)

I had planned to work more on the role management system first, but I need to reconsider a bit of what is and isn't in scope for that before continuing with it.

**Comment by HelmicNewciv** [comment]
Don't know the scope of the role management, but would it include automated ways to add roles?  I know we have a role that we're currently automatically adding to any users that have been in the server for 7 days, but we'd like to restrict it to those who've posted at least X times per day in non-junk channels for X amount of days to filter out idlers.

"Invisible" roles in the form of manually adding users to a channel/category's permissions would be ideal to keep a truly secret channel category.  Goal is to keep a secret community category where users post selfies and personal information, and keep it private and away from harassers.  Not a perfect solution, but it requires an attacker to at least be really familiar with the community and dedicated enough to stay active for a good while.  Most of the hostile folk we get will make their intentions obvious minutes within joining.

That bit's probably out of the scope of the cog, but I'd like to see that functionality somewhere.

**Comment by mikeshardmind** [comment]
@HelmicNewciv I'm up for discussing details on the invisible roles part as part of the cog, including an API for other cogs to say "give this user this invisible role" and just let the cog handle the updating required. 

That isn't a yes or no on doing it (it may come down to rate limits on channel updates, or if not, there will probably be a warning of some sort when using it)

There's also the issue of 6(?) properties of channels not actually being as secret as the client may have you believe. People using things like BetterDiscord* can see the names, topics, ids, position, category, and permissions of every channel, not just the ones they have permissions to. This is done for optimization reasons (I'll try to find the link to a statement on this later) 

*Using client mods like BetterDiscord can get you banned, but the types of people you are describing in your use-case may not care.

As for automated ways to add roles... I've been considering it for a while but creating that criteria in a way manageable from discord commands will be difficult to do in a way which satisfies a more general use case. (That doesn't mean I won't try, just that it is a little lower on the priority list than things with clearer paths to a good implementation.)

I'm definitely open to discussing it more either here or via DM on Discord though, as both of these things can be considered in scope if the implementation can be achieved with any level of sanity.

**Comment by mikeshardmind** [comment]
rss is done, work has been started on the suggestionbox cog (overhauled with more stuff) for v3
